File: app/models/entity/page.go
package entity

import "time"

type PageStatus string
type PageVisibility string

const (
	PageStatusDraft       PageStatus = "draft"
	PageStatusPublished   PageStatus = "published"
	PageStatusUnpublished PageStatus = "unpublished"
	PageStatusScheduled   PageStatus = "scheduled"

	PageVisibilityPublic   PageVisibility = "public"
	PageVisibilityPrivate  PageVisibility = "private"
	PageVisibilityUnlisted PageVisibility = "unlisted"
)

type Page struct {
	ID              int              `json:"id"`
	Title           string           `json:"title"`
	Slug            string           `json:"slug"`
	Content         string           `json:"content"`
	Excerpt         string           `json:"excerpt,omitempty"`
	BannerImageBKey string           `json:"bannerImageBKey,omitempty"`
	Status          PageStatus       `json:"status"`
	Visibility      PageVisibility   `json:"visibility"`
	AllowedRoles    []string         `json:"allowedRoles,omitempty"`
	ParentPageID    *int             `json:"parentPageId,omitempty"`
	AllowComments   bool             `json:"allowComments"`
	AllowReactions  bool             `json:"allowReactions"`
	ShowTOC         bool             `json:"showToc"`
	ScheduledFor    *time.Time       `json:"scheduledFor,omitempty"`
	PublishedAt     *time.Time       `json:"publishedAt,omitempty"`
	CreatedAt       time.Time        `json:"createdAt"`
	UpdatedAt       time.Time        `json:"updatedAt"`
	CreatedBy       *User            `json:"createdBy"`
	UpdatedBy       *User            `json:"updatedBy"`
	Authors         []*User          `json:"authors,omitempty"`
	Topics          []*PageTopic     `json:"topics,omitempty"`
	Tags            []*PageTag       `json:"tags,omitempty"`
	MetaDescription string           `json:"metaDescription,omitempty"`
	CanonicalURL    string           `json:"canonicalUrl,omitempty"`
	CommentsCount   int              `json:"commentsCount"`
	ReactionCounts  []ReactionCounts `json:"reactionCounts,omitempty"`
	EmbeddedPosts   []*Post          `json:"embeddedPosts"`
	CachedAt        *time.Time       `json:"cachedAt,omitempty"`
}

type PageTopic struct {
	ID          int    `json:"id" db:"id"`
	Name        string `json:"name" db:"name"`
	Slug        string `json:"slug" db:"slug"`
	Description string `json:"description,omitempty" db:"description"`
	Color       string `json:"color,omitempty" db:"color"`
}

type PageTag struct {
	ID   int    `json:"id" db:"id"`
	Name string `json:"name" db:"name"`
	Slug string `json:"slug" db:"slug"`
}

type PageDraft struct {
	ID              int       `json:"id"`
	PageID          int       `json:"pageId"`
	UserID          int       `json:"userId"`
	Title           string    `json:"title"`
	Slug            string    `json:"slug"`
	Content         string    `json:"content"`
	Excerpt         string    `json:"excerpt,omitempty"`
	BannerImageBKey string    `json:"bannerImageBKey,omitempty"`
	MetaDescription string    `json:"metaDescription,omitempty"`
	ShowTOC         bool      `json:"showToc"`
	DraftData       string    `json:"draftData"`
	UpdatedAt       time.Time `json:"updatedAt"`
}

File: public/models/page.ts
import { User } from "./identity"
import { Post, ReactionCount } from "./post"

export type PageStatus = "draft" | "published" | "unpublished" | "scheduled"
export type PageVisibility = "public" | "private" | "unlisted"
export type PinnedLocation = "footer" | "subheader"

export interface Page {
  id: number
  title: string
  slug: string
  content: string
  excerpt?: string
  bannerImageBKey?: string
  status: PageStatus
  visibility: PageVisibility
  allowedRoles?: string[]
  parentPageId?: number
  allowComments: boolean
  allowReactions: boolean
  showToc: boolean
  scheduledFor?: string
  publishedAt?: string
  createdAt: string
  updatedAt: string
  createdBy: User
  updatedBy: User
  authors?: User[]
  topics?: PageTopic[]
  tags?: PageTag[]
  metaDescription?: string
  canonicalUrl?: string
  commentsCount: number
  reactionCounts?: ReactionCount[]
  embeddedPosts?: Post[]
  cachedAt?: string
}

export interface PageTopic {
  id: number
  name: string
  slug: string
  description?: string
  color?: string
}

export interface PageTag {
  id: number
  name: string
  slug: string
}

export interface PageDraft {
  id: number
  pageId: number
  title: string
  slug: string
  content: string
  excerpt?: string
  bannerImageBKey?: string
  metaDescription?: string
  showToc: boolean
  updatedAt: string
}

